### PR TITLE
feat: overhaul protocol test generation to use actual client

### PR DESCRIPTION
**Pull request by dayaffe**
## Description of changes
- Modify protocol test client generation to use the generated client rather than construct a set of middlewares in the test.
- Update CRT to 0.31
- Adds ProtocolTestClient() for intercepting request before it is sent. This allows protocol tests to assert against the request.
- Add a StaticEndpointResolver() per SRA which allows customers to specify an endpoint to use rather than rely on endpoint resolution. This is also used for testing to bypass not having a RuleSet in regular protocol tests.
- Adds `checkStreamLengthEqualityIfUnseekableActualStreamHasBeenRead` which takes care of an edge case where an unseekable stream has already been read by the signer so we cannot read its data during test assertions. Instead we can check for that condition and compare stream length to expected length (better than no check)

## Scope

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/SmithyTestUtil/RequestTestUtil/HttpRequestTestBase.swift [hunk 1]
```diff
@@ -378,7 +423,14 @@ open class HttpRequestTestBase: XCTestCase {
...
             )
-            XCTAssertTrue(values.contains(expectedQueryItem.value),
+            let expectedValue = expectedQueryItem.value
+            let valueExists: Bool
+            if let expectedValue = expectedValue {
+                valueExists = values.contains(expectedValue)
+            } else {
+                valueExists = values.isEmpty
+            }
+            XCTAssertTrue(valueExists,
```
Comment (dayaffe) [inline Sources/SmithyTestUtil/RequestTestUtil/HttpRequestTestBase.swift:433]: Test was failing as `nil != []` but these should be equivalent

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/endpoints/EndpointResolverGenerator.kt [hunk 2]
```diff
@@ -62,4 +63,17 @@ class EndpointResolverGenerator(
         writer.write("")
         writer.write("extension DefaultEndpointResolver: EndpointResolver {}")
     }
+
+    private fun renderStaticResolver(writer: SwiftWriter) {
```
Comment (dayaffe) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/endpoints/EndpointResolverGenerator.kt:67]: Static Resolver will be used by tests but are also part of SRA to allow users to bypass endpoint resolution and provide their own endpoint

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt [hunk 3]
```diff
@@ -29,20 +32,33 @@ open class HttpProtocolUnitTestRequestGenerator protected constructor(builder: B
...
+
+        // Normalize the URI
+        val normalizedUri = when {
+            test.uri == "/" -> "/"
+            test.uri.isEmpty() -> ""
+            else -> {
+                val trimmedUri = test.uri.removeSuffix("/")
+                if (!trimmedUri.startsWith('/')) "/$trimmedUri" else trimmedUri
+            }
+        }
```
Comment (dayaffe) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt:51]: Tests were failing due to a missing preceding or ending `/`

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt [hunk 4]
```diff
@@ -29,20 +32,33 @@ open class HttpProtocolUnitTestRequestGenerator protected constructor(builder: B
...
         renderExpectedBlock(test)
         writer.write("")
+        renderClientBlock(test)
         renderOperationBlock(test)
     }
 
     private fun renderExpectedBlock(test: HttpRequestTestCase) {
-        var resolvedHostValue = test.resolvedHost?.let { it } ?: run { "nil" }
+        var resolvedHostValue = if (test.resolvedHost.isPresent && test.resolvedHost.get() != "") test.resolvedHost.get() else "example.com"
+        var hostValue = if (test.host.isPresent && test.host.get() != "") test.host.get() else "example.com"
```
Comment (dayaffe) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt:41]: Provide a host (example.com) to be used during tests that do not specify one

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt [hunk 5]
```diff
@@ -68,8 +84,26 @@ open class HttpProtocolUnitTestRequestGenerator protected constructor(builder: B
         }
     }
 
+    private fun renderClientBlock(test: HttpRequestTestCase) {
```
Comment (dayaffe) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt:87]: This block renders the client call and its config options

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt [hunk 6]
```diff
@@ -68,8 +84,26 @@ open class HttpProtocolUnitTestRequestGenerator protected constructor(builder: B
...
+
+        if (!serviceShape.getTrait(EndpointRuleSetTrait::class.java).isPresent) {
+            val host: String? = test.host.orElse(null)
+            val url: String = "http://${host ?: "example.com"}"
+            writer.write("\nlet config = try await ${clientName}.${clientName}Configuration(endpointResolver: StaticEndpointResolver(endpoint: try \$N(urlString: \$S)))", SmithyHTTPAPITypes.Endpoint, url)
+        } else {
+            writer.write("\nlet config = try await ${clientName}.${clientName}Configuration()")
+        }
+        writer.write("config.region = \"us-west-2\"")
+        writer.write("config.httpClientEngine = ProtocolTestClient()")
```
Comment (dayaffe) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt:99]: Use the ProtocolTestClient() as the httpClientEngine to ensure we get the request back and never send it

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt [hunk 7]
```diff
@@ -68,8 +84,26 @@ open class HttpProtocolUnitTestRequestGenerator protected constructor(builder: B
...
+        writer.write("config.idempotencyTokenGenerator = ProtocolTestIdempotencyTokenGenerator()")
```
Comment (dayaffe) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestRequestGenerator.kt:100]: Use our own token generator to ensure the expected test token will get generated during middleware execution

File: Sources/SmithyTestUtil/RequestTestUtil/HttpRequestTestBase.swift [hunk 8]
```diff
@@ -237,6 +264,17 @@ open class HttpRequestTestBase: XCTestCase {
...
         let actualData = try await actual.readData()
+
+        // Unseekable streams may have already been read by Signer middleware and cannot be read again
+        // Compare stream lengths if ByteStream is a .stream and actualData is nil
+        if checkStreamLengthEqualityIfUnseekableActualStreamHasBeenRead(
+            expected: expected, actual: actual, file: file, line: line
+        ) {
+            // Stream lengths were checked and comparing data will result in failure due to above conditions
+            return
+        }
```
Comment (dayaffe) [inline Sources/SmithyTestUtil/RequestTestUtil/HttpRequestTestBase.swift:275]: There is an edge case where the SignerMiddleware has already read an unseekable stream. In which case by the time all middlewares are executed we cannot re-read the data. In that case let's check if the actual and expected stream lengths are the same.

Note: This comparison should only fail if the actualStream.position == actualStream.length and !actualStream.isSeekable

File: Sources/SmithyTestUtil/ProtocolTestClient.swift [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    public init() {}
+}
+
+public enum TestCheckError: Error {
+    case actual(SdkHttpRequest)
+}
+
+extension ProtocolTestClient: HTTPClient {
+    public func send(request: SdkHttpRequest) async throws -> HttpResponse {
+        throw TestCheckError.actual(request)
```
Comment (dayaffe) [inline Sources/SmithyTestUtil/ProtocolTestClient.swift:21]: throwing an error with the request will allow tests to do checks against the result

File: Sources/ClientRuntime/Networking/Http/Middlewares/SignerMiddleware.swift [hunk 10]
```diff
@@ -69,10 +70,17 @@ extension SignerMiddleware: ApplySigner {
             )
         }
 
+        // Check if CRT should be provided a pre-computed Sha256 SignedBodyValue
+        var updatedSigningProperties = signingProperties
+        let sha256: String? = attributes.get(key: AttributeKey(name: "X-Amz-Content-Sha256"))
+        if let bodyValue = sha256 {
+            updatedSigningProperties.set(key: AttributeKey(name: "SignedBodyValue"), value: sha256)
+        }
```
Comment (dayaffe) [inline Sources/ClientRuntime/Networking/Http/Middlewares/SignerMiddleware.swift:78]: SHA256Middleware if executed will set this attribute to be picked up by the SignerMiddleware and put into the signing properties

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/endpoints/EndpointResolverGenerator.kt [hunk 11]
```diff
@@ -62,4 +63,17 @@ class EndpointResolverGenerator(
...
+        writer.write("")
+        writer.openBlock("public struct StaticEndpointResolver: EndpointResolver {", "}") {
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/endpoints/EndpointResolverGenerator.kt:69]: Can we provide a reusable static endpoint resolver?

The only difference between this endpoint resolver for each service is that the `resolve` method takes that client's `EndpointParams` so we can make that type generic in a reusable static endpoint resolver, then simply use a typedef in each client to provide a specialized endpoint resolver.

**Review by jbelkins — CHANGES_REQUESTED** [review]
See my question about a reusable static endpoint resolver.

Other than that, all is good.

**Review by jbelkins — APPROVED** [review]
